app/src/main/java/org/openobservatory/ooniprobe/activity/MainActivity.java

```diff
@@ -181,55 +191,220 @@ private void requestNotificationPermission() {
+                            .withNegativeButton("")
+                            .withPositiveButton(getString(R.string.Modal_OK))
+                            .build().show(getSupportFragmentManager(), null);
+                }
             }
         }
     }
 
+    /**
+     * The task to fetch the descriptor from the link is completed.
+     * <p>
+     * This method is called when the `fetchDescriptorFromRunId` task is completed.
+     * The `descriptorResponse` is the result of the task.
+     * If the task is successful, the `descriptorResponse` is the descriptor.
+     * Otherwise, the `descriptorResponse` is null.
+     * <p>
+     * If the `descriptorResponse` is not null, start the `AddDescriptorActivity`.
+     * Otherwise, show an error message.
+     *
+     * @param descriptorResponse The result of the task.
+     * @return null.
+     */
+    private Unit fetchDescriptorComplete(OONIRunDescriptor descriptorResponse) {
+        if (descriptorResponse != null) {
+            startActivity(AddDescriptorActivity.newIntent(this, descriptorResponse));
+        } else {
+            // TODO(aanorbel): Provide a better error message.
+            Snackbar.make(binding.getRoot(), R.string.Modal_Error, Snackbar.LENGTH_LONG)
+                    .setAnchorView(binding.bottomNavigation) // NOTE:To avoid the `snackbar` from covering the bottom navigation.
+                    .show();
+        }
+        removeProgressFragment();
+        return Unit.INSTANCE;
+
+    }
+
+    /**
+     * Display the progress fragment.
+     * <p>
+     * The progress fragment is used to display the progress of the task.
+     * e.g. Fetching the descriptor from the link.
+     *
+     * @param executor The executor that will be used to execute the task.
+     */
+    private void displayAddLinkProgressFragment(TaskExecutor executor) {
+        binding.dynamicProgressFragment.setVisibility(View.VISIBLE);
+        getSupportFragmentManager().beginTransaction()
+                .add(
+                        R.id.dynamic_progress_fragment,
+                        OONIRunDynamicProgressBar.newInstance(ProgressType.ADD_LINK, new OnActionListener() {
+                            @Override
+                            public void onActionButtonCLicked() {
+                                executor.cancelTask();
+                                removeProgressFragment();
+                            }
+
+                            @Override
+                            public void onCloseButtonClicked() {
+                                removeProgressFragment();
+                            }
+                        }),
+                        OONIRunDynamicProgressBar.getTAG()
+                ).commit();
+    }
+
+    /**
+     * Extracts the run id from the provided Uri.
+     * The run id can be in two different formats:
+     * <p>
+     * 1. ooni://runv2/#link_id
+     * 2. https://run.test.ooni.org/v2/#link_id
```

Comment:
```suggestion
     * 1. ooni://runv2/link_id
     * 2. https://run.test.ooni.org/v2/link_id
```

We cannot use `#` because it has a specific semantics in HTTP URLs.